menas/src/main/scala/za/co/absa/enceladus/rest/controllers/HDFSController.scala

```diff
@@ -34,28 +36,33 @@ class HDFSController @Autowired() (fs: FileSystem) {
   private val logger = LoggerFactory.getLogger(this.getClass)
   
   @PostMapping(path = Array("/list"))
-  def getHDFSFolder(@RequestBody path: String): CompletableFuture[HDFSFolder] = {
+  def getHDFSFolder(@RequestBody path: String): CompletableFuture[ResponseEntity[HDFSFolder]] = {
     
     val p = new Path(path)
-    val res = if (!fs.isDirectory(p)) Future { HDFSFolder(p.toUri().getPath, p.getName, None) }
-    else {
-      Future {
-        val status = fs.listStatus(p)
-        val children = if (status.isEmpty) None else {
-          Some(
-            status.map({ x =>
-              val child = x.getPath
-
-              HDFSFolder(child.toUri().getPath, child.getName,
-                if (fs.listStatus(child).isEmpty) None else Some(Seq(HDFSFolder("", "", None))))
-
-            }).toSeq)
+    if (fs.exists(p)) {
+      val res = if (!fs.isDirectory(p)) {
+        Future.successful(HDFSFolder(p.toUri().getPath, p.getName, None))
+      } else {
+        Future {
+          val status = fs.listStatus(p)
+          val children = if (status.isEmpty) None else {
+            Some(
+              status.map({ x =>
+                val child = x.getPath
+
+                HDFSFolder(child.toUri().getPath, child.getName,
+                  if (fs.listStatus(child).isEmpty) None else Some(Seq(HDFSFolder("", "", None))))
+
+              }).toSeq)
+          }
+          HDFSFolder(path, p.getName, children)
         }
-        HDFSFolder(path, p.getName, children)
       }
-    }
 
-    res
+      res.map(result => ResponseEntity.ok(result))
+    } else {
+      Future.successful(ResponseEntity.notFound().build[HDFSFolder]())
```

Comment:
Isn't this called a guard clause or something like that. Basically, you have a huge if and then a small else. What you should do is have an `if not` on top with an early `return`, a one-liner, and rest of the method doesn't need any `else` or things like that.

Reply:
I would agree with you if this was java, but returns are antithetical to scala. Speaking as someone who's had to disentangle exactly this sort of early return mess, I'd rather avoid them entirely than deal with unexpected (`non-local returns`)[https://tpolecat.github.io/2014/05/09/return.html].
However, it's probably a sign that a bit of refactoring is needed here, so thanks for pointing it out.